packages/pigeon/lib/dart_generator.dart

```diff
@@ -500,48 +512,48 @@ void generateDart(DartOptions opt, Root root, StringSink sink) {
       indent.write('Object encode() ');
       indent.scoped('{', '}', () {
         indent.writeln(
-          'final Map<Object?, Object?> pigeonMap = <Object?, Object?>{};',
+          'final List<Object?> pigeonList = <Object?>[];',
         );
         for (final NamedType field in klass.fields) {
-          indent.write("pigeonMap['${field.name}'] = ");
+          indent.write('pigeonList.add(');
           final String conditional = field.type.isNullable ? '?' : '';
           if (customClassNames.contains(field.type.baseName)) {
             indent.addln(
-              '${field.name}$conditional.encode();',
+              '${field.name}$conditional.encode());',
             );
           } else if (customEnumNames.contains(field.type.baseName)) {
             indent.addln(
-              '${field.name}$conditional.index;',
+              '${field.name}$conditional.index);',
             );
           } else {
-            indent.addln('${field.name};');
+            indent.addln('${field.name});');
           }
         }
-        indent.writeln('return pigeonMap;');
+        indent.writeln('return pigeonList;');
       });
     }
 
     void writeDecode() {
-      void writeValueDecode(NamedType field) {
+      void writeValueDecode(NamedType field, int index) {
         if (customClassNames.contains(field.type.baseName)) {
           final String nonNullValue =
-              "${field.type.baseName}.decode(pigeonMap['${field.name}']!)";
+              '${field.type.baseName}.decode(result[$index]! as List<Object?>)';
```

Comment:
Let's extract `result[$index]` as a local variable (in the generator, not the generated code) instead of repeating it seven times in this method. Then the next time we change this logic it'll be easier :)

Reply:
I never know when to extract string segments like this. On one hand, it's faster to change it later, for sure. On the other hand, removing string literals for variables makes reading the generators a lot more difficult. There's a trade-off there.